discovery/kubernetes/ingress.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+	"golang.org/x/net/context"
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Ingress implements discovery of Kubernetes ingresss.
+type Ingress struct {
+	logger   log.Logger
+	informer cache.SharedInformer
+	store    cache.Store
+}
+
+// NewIngress returns a new ingress discovery.
+func NewIngress(l log.Logger, inf cache.SharedInformer) *Ingress {
+	return &Ingress{logger: l, informer: inf, store: inf.GetStore()}
+}
+
+// Run implements the TargetProvider interface.
+func (s *Ingress) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// Send full initial set of pod targets.
+	var initial []*config.TargetGroup
+	for _, o := range s.store.List() {
+		tg := s.buildIngress(o.(*v1beta1.Ingress))
```

Comment:
Is this interface conversion always safe? Shall we guard against a panic?

Reply:
It's an odd design, but the cache actually makes sure that it's the right type before inserting it into cache, so yes it's actually safe here to do this, otherwise something went utterly wrong and all of Kubernetes is broken :slightly_smiling_face:.

Reply:
Right, also the same in other parts of the k8s discovery.